### PR TITLE
Bug 1839158 - Ignore invalid credit card save requests

### PR DESCRIPTION
**Pull request by MatthewTighe**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1839158

### CODE REVIEW

**Comment by zmckenney** [comment]
Just thinking about where this validation should live and the ability to surface the issue to a user, would this be better placed in the `DefaultCreditCardValidationDelegate` and inside of `shouldCreateOrUpdate` we could add a result for `Result.InvalidEntry`? This would place the validation in the validating delegate as well as allow for UI updates to the user easily via the `when` statement for `updateUI` in `CreditCardSaveDialogFragment`. 

Thinking something along these lines:
```diff --git a/android-components/components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt b/android-components/components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt
index 7a50ac458b..f35e5b6295 100644
--- a/android-components/components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt
+++ b/android-components/components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt
@@ -451,6 +451,11 @@ interface CreditCardValidationDelegate {
          * can be used to update its information.
          */
         data class CanBeUpdated(val foundCreditCard: CreditCard) : Result()
+
+        /**
+         * Indicates the [CreditCard] does not include all required fields
+         */
+        object InvalidEntry : Result()
     }
 
     /**
diff --git a/android-components/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/creditcard/CreditCardSaveDialogFragment.kt b/android-components/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/creditcard/CreditCardSaveDialogFragment.kt
index 428c48fc66..1c2c3a26c8 100644
--- a/android-components/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/creditcard/CreditCardSaveDialogFragment.kt
+++ b/android-components/components/feature/prompts/src/main/java/mozilla/components/feature/prompts/creditcard/CreditCardSaveDialogFragment.kt
@@ -31,6 +31,7 @@ import mozilla.components.feature.prompts.dialog.KEY_SESSION_ID
 import mozilla.components.feature.prompts.dialog.KEY_SHOULD_DISMISS_ON_LOAD
 import mozilla.components.feature.prompts.dialog.PromptDialogFragment
 import mozilla.components.feature.prompts.facts.emitCreditCardAutofillCreatedFact
+import mozilla.components.feature.prompts.facts.emitCreditCardAutofillDismissedFact
 import mozilla.components.feature.prompts.facts.emitCreditCardAutofillUpdatedFact
 import mozilla.components.support.ktx.android.view.toScope
 import mozilla.components.support.utils.creditCardIssuerNetwork
@@ -118,6 +119,9 @@ internal class CreditCardSaveDialogFragment : PromptDialogFragment() {
             is Result.CanBeUpdated -> {
                 emitCreditCardAutofillUpdatedFact()
             }
+            is Result.InvalidEntry -> {
+                emitCreditCardAutofillDismissedFact()
+            }
         }
     }
 
@@ -151,6 +155,13 @@ internal class CreditCardSaveDialogFragment : PromptDialogFragment() {
                     confirmButtonText = requireContext().getString(R.string.mozac_feature_prompt_update_confirmation),
                     showMessageBody = false,
                 )
+                is Result.InvalidEntry -> setViewText(
+                    view = view,
+                    header = "This CC was missing required fields",
+                    cancelButtonText = "OK",
+                    confirmButtonText = "",
+                    showMessageBody = false
+                )
             }
         }
     }
diff --git a/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/DefaultCreditCardValidationDelegate.kt b/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/DefaultCreditCardValidationDelegate.kt
index d3e6122fcb..2092a407ff 100644
--- a/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/DefaultCreditCardValidationDelegate.kt
+++ b/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/DefaultCreditCardValidationDelegate.kt
@@ -26,6 +26,7 @@ class DefaultCreditCardValidationDelegate(
 
     override suspend fun shouldCreateOrUpdate(creditCard: CreditCardEntry): Result =
         withContext(coroutineContext) {
+            if (creditCard.number.isEmpty() || creditCard.expiryDate.isEmpty()) return@withContext Result.InvalidEntry
             val creditCards = storage.value.getAllCreditCards()
 
             val foundCreditCard = if (creditCards.isEmpty()) {
diff --git a/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/GeckoCreditCardsAddressesStorageDelegate.kt b/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/GeckoCreditCardsAddressesStorageDelegate.kt
index cb2da4b868..a15d2512eb 100644
--- a/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/GeckoCreditCardsAddressesStorageDelegate.kt
+++ b/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/GeckoCreditCardsAddressesStorageDelegate.kt
@@ -99,6 +99,9 @@ class GeckoCreditCardsAddressesStorageDelegate(
                         ),
                     )
                 }
+                is CreditCardValidationDelegate.Result.InvalidEntry -> {
+                    /** No Op **/
+                }
             }
         }
     }
```

**Comment by MatthewTighe** [comment]
> Just thinking about where this validation should live and the ability to surface the issue to a user, would this be better placed in the DefaultCreditCardValidationDelegate and inside of shouldCreateOrUpdate we could add a result for Result.InvalidEntry?

That was [my intuition as well](https://github.com/mozilla-mobile/firefox-android/pull/3052/files). This fix would at least reduce cycle time by not requiring UX or product input on strings etc

File: android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/GeckoCreditCardsAddressesStorageDelegate.kt [hunk 1]
```diff
@@ -72,6 +72,7 @@ class GeckoCreditCardsAddressesStorageDelegate(
         }
 
     override suspend fun onCreditCardSave(creditCard: CreditCardEntry) {
+        if (!creditCard.isValid) return
```
Comment (gabrielluong) [inline android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/GeckoCreditCardsAddressesStorageDelegate.kt:75]: Let's add a new line after this.

**Review by gabrielluong — APPROVED** [review]

**Comment by zmckenney** [comment]
> > Just thinking about where this validation should live and the ability to surface the issue to a user, would this be better placed in the DefaultCreditCardValidationDelegate and inside of shouldCreateOrUpdate we could add a result for Result.InvalidEntry?
> 
> That was [my intuition as well](https://github.com/mozilla-mobile/firefox-android/pull/3052/files). This fix would at least reduce cycle time by not requiring UX or product input on strings etc

Couldn't we just leave the invalid entry as NO-OP with comment in the location of the UI showing? It would probably even be beneficial to see this as something not handled by the UI in this case. Either way, this is supposed to be a quick fix so I understand if this is the way we want to do it for now.

**Comment by gabrielluong** [comment]
> > Just thinking about where this validation should live and the ability to surface the issue to a user, would this be better placed in the DefaultCreditCardValidationDelegate and inside of shouldCreateOrUpdate we could add a result for Result.InvalidEntry?
> 
> That was [my intuition as well](https://github.com/mozilla-mobile/firefox-android/pull/3052/files). This fix would at least reduce cycle time by not requiring UX or product input on strings etc

I think this ultimately requires some UX/Product decisions. The current issue is that the save credit card dialog is displayed as long as there's a card number (w/o expiry date) when a payment form is submitted. There's 2 options:
1. We do not display this dialog if we do not have a complete credit card entry. 
2. We display the prompt but add handling of missing entries. Either allow the user to fill in missing inputs or cancel saving gracefully.

I would suggest filing a new bug to investigate the options here. Currently, I am more for option 1 because our autofill typically doesn't fill in the expiry date, which will make it easy to hit the update credit card dialog if the user submits the payment form without being aware that certain fields are missing from the autofill.

**Comment by MatthewTighe** [comment]
>  Either way, this is supposed to be a quick fix so I understand if this is the way we want to do it for now.

Yeah, in the interest of quickly addressing the crash I am going to land as is but I filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1846631) to track future improvements.

**Comment by MatthewTighe** [comment]
@mergify backport releases_v116

**Comment by MatthewTighe** [comment]
@mergify backport releases_v117

**Pull request merged**